cloud/openstack/os_port.py

```diff
@@ -0,0 +1,382 @@
+            return True
+    for key in compare_dict:
+        if module.params[key] is not None and cmp(module.params[key],
+                                                  port[key]) != 0:
+            return True
+    for key in compare_comma_separated_list:
+        if module.params[key] is not None and (set(module.params[key] if
+                    type(module.params[key]) == list else
+                    module.params[key].split(',')) != set(port[key])):
+            return True
+
+    # NOTE: if port was created or updated with 'no_security_groups=True',
+    # subsequent updates without 'no_security_groups' flag or
+    # 'no_security_groups=False' and no specified 'security_groups', will not
+    # result in an update to the port where the default security group is
+    # applied.
+    if module.params['no_security_groups'] and port['security_groups'] != []:
+        return True
+
+    if module.params['fixed_ips'] is not None:
+        for item in module.params['fixed_ips']:
+            if 'ip_address' in item:
+                # if ip_address in request does not match any in existing port,
+                # update is required.
+                if not any(match['ip_address'] == item['ip_address']
+                           for match in port['fixed_ips']):
+                    return True
+            if 'subnet_id' in item:
+                return True
+        for item in port['fixed_ips']:
+            # if ip_address in existing port does not match any in request,
+            # update is required.
+            if not any(match.get('ip_address') == item['ip_address']
+                       for match in module.params['fixed_ips']):
+                return True
+
+    return False
+
+
+def _system_state_change(module, port):
+    state = module.params['state']
+    if state == 'present':
+        if not port:
+            return True
+        return _needs_update(module, port)
+    if state == 'absent' and port:
+        return True
+    return False
+
+
+def _compose_port_args(module):
+    port_kwargs = {}
+    optional_parameters = ['name',
+                           'fixed_ips',
+                           'admin_state_up',
+                           'mac_address',
+                           'allowed_address_pairs',
+                           'extra_dhcp_opt',
+                           'device_owner',
+                           'device_id']
+    for optional_param in optional_parameters:
+        if module.params[optional_param] is not None:
+            port_kwargs[optional_param] = module.params[optional_param]
+
+    if module.params['no_security_groups']:
+        port_kwargs['security_groups'] = []
+    elif module.params['security_groups']:
+        if type(module.params['security_groups']) == str:
+            port_kwargs['security_groups'] = module.params[
+                'security_groups'].split(',')
+        else:
+            port_kwargs['security_groups'] = module.params['security_groups']
+
```

Comment:
You *could* add this, but it might slow the module quite a bit. In order to support passing name or ID to get_security_group(), a list of *all* security groups is returned and then searched for a match on name or ID. This would be done for each security group passed in to this module. This isn't very efficient, and something we hope to change in later versions of shade.

I would suggest leaving it as is, and maybe we could add support for sec group names later.